Add base dcm class

Illviljan opened the pull request:
Example attempt to align return types for values.

The API is inspired by [numpy](https://github.com/numpy/numpy) and [xarray](https://github.com/pydata/xarray).

High level ideas:
* Inherit/use only a single dcm class, reduces typo risks etc, Risk of getting too much clutter.
* Follow typical conventions popular in data analysis tools like [numpy](https://github.com/numpy/numpy) /[xarray ](https://github.com/pydata/xarray)
  * Use self.shape instead of x_dimension y_dimension. -> Following numpy.array.shape standard.
  * Use list of list for WERT values. -> Easy to initialize numpy.array then.
  * Use tuple of lists for axis values.  -> Matching order of numpy.array.shape-tuples
  * Move remaining metadata to a self.attrs dict. -> In similar fashion as xarray/pandas
* A lot of type hints to clarify intention of methods and attributes.


Only maps works in this form for now. I want some discussion before going all in on this.

References:
https://www.etas.com/download-center-files/products_ASCET_Software_Products/TechNote_DCM_File_Formats.pdf

Illviljan committed:
Create utils.py

Illviljan committed:
Update utils.py

Illviljan committed:
Update utils.py

Illviljan committed:
Update dcm_reader.py

Illviljan committed:
Update dcm_characteristic_map.py

Illviljan committed:
Update dcm_fixed_characteristic_map.py

Illviljan committed:
Update dcm_group_characteristic_map.py

Illviljan committed:
Merge remote-tracking branch 'upstream/main' into align_classes_numpy_like

Illviljan committed:
Update dcm_reader.py

Illviljan committed:
Update dcm_reader.py

Illviljan committed:
Update dcm_reader.py

dkuschmierz commented:
Wouldn't it be better to directly use numpy within the project instead of generating compatible interfaces?

Illviljan commented:
That's ok as well. I was just thinking dcmReader didn't wan't to have numpy as a dependency. Should numpy be a dependency?

Illviljan committed:
better dcm format printer

Illviljan committed:
Fixes for base class

Illviljan committed:
Use similar code paths for parsing the blocks.

Illviljan committed:
Use dcm base class

Illviljan committed:
Ok I give up on doing this without numpy

FESTWERTEBLOCK blockParameter2D 4 @ 2 was the last straw. How are you supposed to keep track of dimensions without the st/y and handle if they're split because linelength limits?

Going with numpy will deal with all the annoying error handling and once the values are all parsed in a big 1d array then do a simple reshape.

Illviljan committed:
Create py.typed

Illviljan committed:
remove duplicated parse code

Illviljan committed:
use ndarray's, generalize print format

Illviljan committed:
move to elements

Illviljan committed:
Don't use os.linesep it's different and messes with tests.

Illviljan committed:
Rename Tests to test_dcmreader so pytest finds the file

Illviljan committed:
use join for multiple comments, removes trailing \n

Illviljan committed:
y/x_mapping to name_y/x

Illviljan committed:
block_type to element_syntax, matches better with the technical notes

Illviljan committed:
Update tests

Illviljan committed:
Update dcm_reader.py

Illviljan committed:
Move to pytest and parametrize

Illviljan committed:
Various fixes
